File: scripts/route_smoke.py
#!/usr/bin/env python3
"""
Route Smoke Test — Playwright

Opens each /command and /workspace/:companyId/:periodYear/* route against the
running dev/preview server and asserts:
  - Final HTTP status 200 with no redirect loop
  - Non-empty <title>
  - A heading/main element renders
  - companyId + periodYear survive to the final URL on workspace routes
  - No uncaught page errors or console errors

Auth-gated routes are SKIPPED (not failed) when LOVABLE_BROWSER_AUTH_STATUS is
not 'injected', so this script is safe to run unauthenticated in CI.

Env:
  BASE_URL           default http://localhost:8080
  SMOKE_COMPANY_ID   default 00000000-0000-0000-0000-000000000001
  SMOKE_PERIOD_YEAR  default 2025

Exit code 0 = all PASS/SKIP. Exit 1 = one or more FAIL.
"""

import asyncio
import json
import os
import sys
from urllib.parse import urlparse

from playwright.async_api import async_playwright

BASE_URL = os.environ.get("BASE_URL", "http://localhost:8080")
COMPANY_ID = os.environ.get("SMOKE_COMPANY_ID", "00000000-0000-0000-0000-000000000001")
PERIOD_YEAR = os.environ.get("SMOKE_PERIOD_YEAR", "2025")
AUTH_STATUS = os.environ.get("LOVABLE_BROWSER_AUTH_STATUS", "signed_out")
AUTHENTICATED = AUTH_STATUS == "injected"
MAX_REDIRECTS = 5

WORKSPACE_STAGES = [
    "",  # overview
    "prepare",
    "reconcile",
    "statements",
    "tax",
    "compliance",
    "filing",
    "monitor",
]

ROUTES = [
    {"path": "/command", "requires_auth": False, "check_params": False},
    *[
        {
            "path": f"/workspace/{COMPANY_ID}/{PERIOD_YEAR}"
            + (f"/{stage}" if stage else ""),
            "requires_auth": True,
            "check_params": True,
        }
        for stage in WORKSPACE_STAGES
    ],
]

GREEN = "\033[32m"
RED = "\033[31m"
YELLOW = "\033[33m"
RESET = "\033[0m"


def color(status: str) -> str:
    return {"PASS": GREEN, "FAIL": RED, "SKIP": YELLOW}.get(status, "") + status + RESET


async def restore_session(context, page):
    """Restore Supabase session from env if injected."""
    storage_key = os.environ.get("LOVABLE_BROWSER_SUPABASE_STORAGE_KEY")
    session_json = os.environ.get("LOVABLE_BROWSER_SUPABASE_SESSION_JSON")
    cookies_json = os.environ.get("LOVABLE_BROWSER_SUPABASE_COOKIES_JSON")
    if cookies_json:
        cookies = json.loads(cookies_json)
        for c in cookies:
            c["url"] = BASE_URL
        await context.add_cookies(cookies)
    await page.goto(BASE_URL, wait_until="domcontentloaded")
    if storage_key and session_json:
        await page.evaluate(
            f"window.localStorage.setItem({json.dumps(storage_key)}, {json.dumps(session_json)})"
        )


async def test_route(browser, route):
    path = route["path"]
    url = f"{BASE_URL}{path}"

    if route["requires_auth"] and not AUTHENTICATED:
        return {"path": path, "status": "SKIP", "detail": "auth not injected"}

    context = await browser.new_context(viewport={"width": 1280, "height": 900})
    page = await context.new_page()

    page_errors: list[str] = []
    console_errors: list[str] = []
    redirects: list[str] = []

    page.on("pageerror", lambda err: page_errors.append(str(err)))
    page.on(
        "console",
        lambda msg: console_errors.append(msg.text) if msg.type == "error" else None,
    )
    page.on(
        "response",
        lambda resp: redirects.append(resp.url)
        if 300 <= resp.status < 400 and resp.url.startswith(BASE_URL)
        else None,
    )

    try:
        if AUTHENTICATED:
            await restore_session(context, page)

        resp = await page.goto(url, wait_until="domcontentloaded", timeout=15000)
        if resp is None:
            return {"path": path, "status": "FAIL", "detail": "no response"}
        if resp.status != 200:
            return {"path": path, "status": "FAIL", "detail": f"HTTP {resp.status}"}
        if len(redirects) > MAX_REDIRECTS:
            return {
                "path": path,
                "status": "FAIL",
                "detail": f"redirect loop ({len(redirects)})",
            }

        try:
            await page.wait_for_load_state("networkidle", timeout=6000)
        except Exception:
            pass

        title = (await page.title()).strip()
        if not title:
            return {"path": path, "status": "FAIL", "detail": "empty <title>"}

        final_url = page.url
        final_path = urlparse(final_url).path

        if route["requires_auth"] and "/auth" in final_path:
            return {"path": path, "status": "SKIP", "detail": "redirected to /auth"}

        if route["check_params"]:
            if COMPANY_ID not in final_path:
                return {
                    "path": path,
                    "status": "FAIL",
                    "detail": f"companyId lost → {final_path}",
                }
            if f"/{PERIOD_YEAR}" not in final_path:
                return {
                    "path": path,
                    "status": "FAIL",
                    "detail": f"periodYear lost → {final_path}",
                }

        has_content = await page.evaluate(
            "(document.body && (document.body.innerText || '').trim().length > 0)"
        )
        if not has_content:
            return {"path": path, "status": "FAIL", "detail": "no rendered content"}

        errs = page_errors + console_errors
        if errs:
            return {
                "path": path,
                "status": "FAIL",
                "detail": "errors: " + " | ".join(errs[:2]),
            }

        return {"path": path, "status": "PASS", "detail": title[:60]}
    except Exception as e:
        return {"path": path, "status": "FAIL", "detail": str(e)[:200]}
    finally:
        await context.close()


async def main():
    print(f"\nRoute smoke test → {BASE_URL}")
    print(
        f"Auth: {'INJECTED' if AUTHENTICATED else 'SIGNED OUT (workspace routes will SKIP)'}\n"
    )

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        results = []
        for route in ROUTES:
            r = await test_route(browser, route)
            results.append(r)
            print(f"  {color(r['status']):>18}  {r['path']:<60}  {r.get('detail', '')}")
        await browser.close()

    passed = sum(1 for r in results if r["status"] == "PASS")
    failed = sum(1 for r in results if r["status"] == "FAIL")
    skipped = sum(1 for r in results if r["status"] == "SKIP")

    print(f"\n{passed} passed · {failed} failed · {skipped} skipped")
    print("VERDICT: " + ("CLEAN" if failed == 0 else "FAIL"))
    sys.exit(0 if failed == 0 else 1)


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/route_smoke.mjs
#!/usr/bin/env node
/**
 * Route Smoke Test — Playwright
 *
 * Opens each /command and /workspace/:companyId/:periodYear/* route against the
 * running dev/preview server and asserts:
 *   - HTTP-level: no redirect loop, final status 200
 *   - DOM: <title> present and non-empty; <h1>/main heading renders
 *   - Params: companyId + periodYear survive to the final URL (workspace routes)
 *   - No uncaught console errors
 *
 * Auth-gated routes are skipped when LOVABLE_BROWSER_AUTH_STATUS !== 'injected'
 * and reported as SKIPPED (not FAIL) so the script is safe to run unauthenticated.
 *
 * Env:
 *   BASE_URL          default http://localhost:8080
 *   SMOKE_COMPANY_ID  default 00000000-0000-0000-0000-000000000001
 *   SMOKE_PERIOD_YEAR default 2025
 *
 * Exit code 0 = all PASS/SKIP. Exit code 1 = one or more FAIL.
 */

import { chromium } from 'playwright'

const BASE_URL     = process.env.BASE_URL          ?? 'http://localhost:8080'
const COMPANY_ID   = process.env.SMOKE_COMPANY_ID  ?? '00000000-0000-0000-0000-000000000001'
const PERIOD_YEAR  = process.env.SMOKE_PERIOD_YEAR ?? '2025'
const AUTH_STATUS  = process.env.LOVABLE_BROWSER_AUTH_STATUS ?? 'signed_out'
const AUTHENTICATED = AUTH_STATUS === 'injected'

const WORKSPACE_STAGES = [
  '',            // overview (index route)
  'prepare',
  'reconcile',
  'statements',
  'tax',
  'compliance',
  'filing',
  'monitor',
]

const ROUTES = [
  { path: '/command', requiresAuth: false, checkParams: false },
  ...WORKSPACE_STAGES.map((stage) => ({
    path: `/workspace/${COMPANY_ID}/${PERIOD_YEAR}${stage ? `/${stage}` : ''}`,
    requiresAuth: true,
    checkParams: true,
  })),
]

const MAX_REDIRECTS = 5

function fmt(status, label, detail = '') {
  const color = status === 'PASS' ? '\x1b[32m' : status === 'FAIL' ? '\x1b[31m' : '\x1b[33m'
  return `${color}${status}\x1b[0m  ${label}${detail ? '  ' + detail : ''}`
}

async function testRoute(browser, route) {
  const { path, requiresAuth, checkParams } = route
  const url = `${BASE_URL}${path}`

  if (requiresAuth && !AUTHENTICATED) {
    return { path, status: 'SKIP', reason: 'auth not injected' }
  }

  const context = await browser.newContext({ viewport: { width: 1280, height: 900 } })
  const page = await context.newPage()

  const consoleErrors = []
  page.on('pageerror', (err) => consoleErrors.push(err.message))
  page.on('console', (msg) => {
    if (msg.type() === 'error') consoleErrors.push(msg.text())
  })

  const redirects = []
  page.on('response', (resp) => {
    if (resp.status() >= 300 && resp.status() < 400 && resp.url().startsWith(BASE_URL)) {
      redirects.push(resp.url())
    }
  })

  try {
    const resp = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 15000 })
    if (!resp) return { path, status: 'FAIL', reason: 'no response' }
    if (resp.status() !== 200) {
      return { path, status: 'FAIL', reason: `HTTP ${resp.status()}` }
    }

    if (redirects.length > MAX_REDIRECTS) {
      return { path, status: 'FAIL', reason: `redirect loop (${redirects.length})` }
    }

    // Wait for React render
    await page.waitForLoadState('networkidle', { timeout: 8000 }).catch(() => {})

    const title = (await page.title()).trim()
    if (!title) return { path, status: 'FAIL', reason: 'empty <title>' }

    const finalUrl = page.url()
    // Auth redirect: workspace routes redirect to /auth when signed out
    if (requiresAuth && finalUrl.includes('/auth')) {
      return { path, status: 'SKIP', reason: 'redirected to /auth' }
    }

    if (checkParams) {
      const u = new URL(finalUrl)
      if (!u.pathname.includes(COMPANY_ID)) {
        return { path, status: 'FAIL', reason: `companyId lost — final ${u.pathname}` }
      }
      if (!u.pathname.includes(`/${PERIOD_YEAR}`)) {
        return { path, status: 'FAIL', reason: `periodYear lost — final ${u.pathname}` }
      }
    }

    // Assert something rendered (main region or heading)
    const hasContent = await page.evaluate(() => {
      const main = document.querySelector('main, [role="main"], h1, h2')
      return !!main && (main.textContent ?? '').trim().length > 0
    })
    if (!hasContent) return { path, status: 'FAIL', reason: 'no rendered heading/main' }

    if (consoleErrors.length > 0) {
      return {
        path,
        status: 'FAIL',
        reason: `console errors: ${consoleErrors.slice(0, 2).join(' | ')}`,
      }
    }

    return { path, status: 'PASS', title, finalUrl }
  } catch (err) {
    return { path, status: 'FAIL', reason: err.message }
  } finally {
    await context.close()
  }
}

async function main() {
  console.log(`\nRoute smoke test → ${BASE_URL}`)
  console.log(`Auth: ${AUTHENTICATED ? 'INJECTED' : 'SIGNED OUT (workspace routes will SKIP)'}\n`)

  const browser = await chromium.launch({ headless: true })
  const results = []
  for (const route of ROUTES) {
    const r = await testRoute(browser, route)
    results.push(r)
    console.log(fmt(r.status, r.path, r.reason ?? r.title ?? ''))
  }
  await browser.close()

  const pass = results.filter((r) => r.status === 'PASS').length
  const fail = results.filter((r) => r.status === 'FAIL').length
  const skip = results.filter((r) => r.status === 'SKIP').length

  console.log(`\n${pass} passed · ${fail} failed · ${skip} skipped`)
  console.log(fail === 0 ? 'VERDICT: CLEAN' : 'VERDICT: FAIL')
  process.exit(fail === 0 ? 0 : 1)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})